### ISSUE TITLE
-p but keep the processes alive

### ISSUE THREAD

**Issue by paddor** [erikbrinkman/xstream#1]
I found this project while looking for an xargs-like tool but for STDIN. It looks promising but for example -p10 spawns way more than 10 processes.

I wish there was an option to just start X number of processes which will be kept alive, fed line-by-line in round-robin fashion, then sent EOF so they all terminate. That would allow spread out work across multiple filters Unix-style efficiently.

Using something like ZMQ or NNG, one could then re-combine the results (map->reduce). But xstream would be the tool at the beginning of the pipeline, splitting the large input file.

**Comment by erikbrinkman** [comment]
Yeah, that idea makes sense. I hadn't touched this utility in a while, and I've gotten better at rust in the mean time.

I implemented exactly what you specified. Now if you execute `xstream -p10 -r -- ...` it'll spawn up to 10 processes, and then keep rotating them, almost as if you just had 10 inputs that skipped every 10 lines.

One thing to note is that I have still basically copped out on the output, meaning each process just tries to fight over control stdout. For simple things this is probably fine, but if your output is complex, you'll likely see race-conditions in your output.

In principle, this could be remedied by capturing the stdout to a buffer and then dumping that out to a writer, but it means potentially having to keep each processes stdout in memory. Also, with the round robin approach, there's no way to know that process has finished dumping the output for one process. We could apply the same methodology, e.g. have a delimiter, and buffer each process by the delimiter. At this point there's also the question ordering, and verifying the child processes enforce that contract.

All of these questions are answerable, but outside the scope of why I wrote this, so if you run into output race conditions, I'd love to have your thoughts on how to handle.
